### PR TITLE
[RFR] Introduce Form Components

### LINKED ISSUE #216: [RFC] Intermediate layout component for `Edit`, `Create`, and `Show` views
## Current status

Currently, the list view is composed of: 

- a data fetching component (`<List>`)
- a layout component (`<Datagrid>`)
- field components to configure the columns (`<TextField>` and others).

```js
const PostList = ({ props }) =>
  <List {...props} defaultSort={{ field: 'created_at', order: 'DESC' }}>
    <Datagrid rowStyle={(record) => ({ backgroudColor: !record.published_at ? 'red' : 'white' })}>
      <TextField source="title" />
    </Datagrid>
    </List>;
```

In contrast, the edit view is composed of:

- a data fetching component, also managing the layout (`<Edit>`)
- input components to configure the form inputs (`<TextInput>` and others).

```js
const PostEdit = ({ props }) =>
  <Edit validation={validationFunction}>
    <TextInput source="title" />
  </Edit>
```

As a consequence, implementing a custom form layout (tabbed, wizard, etc) is hard, and requires specializing the `<Edit> component.

## Suggestion

I suggest to make the API consistent, and require a form layout component between the `<Edit>` and `<Input>` components:

```js
const PostEdit = ({ props }) =>
  <Edit validation={validationFunction}>
    <SimpleForm>
      <TextInput source="title" />
    </SimpleForm>
  </Edit>
```

Alternatives to `<SimpleForm>` could be `<TabbedForm>`, `<WizardForm>`, or whetever layout component the developer feels like using.

## Impact

Writing an Edit view is slightly more verbose. But we had the same remark when [moving the layout out of `<List>`](https://github.com/marmelab/admin-on-rest/pull/22), and it hasn't created any real problem.

This is a major BC break, but it's the right time to do it since [the `next` version already breaks BC](https://github.com/marmelab/admin-on-rest/blob/6aa7e158aaa8e1bab43a44fba156618201bc545c/CHANGELOG.md).

What do you think?

### PR DESCRIPTION
**Pull request by fzaninotto**
- [x] [BC Break] Introduce `<SimpleForm>` component between `<Edit>`/`<Create>` and input components
- [x] Move `validation` and `defaultValue` props to form component
- [x] Add `<FormField>` component to allow reuse of the `addLabel`/`addField` decoration logic
- [x] Introduce `<TabbedForm>` component as an example of alternative form layout
- [x] Add documentation
- [x] Fix `<SavedButton>` in case of invalid form (was spinning forever)
- [x] Fix post creation on example (was lost in infinite loop because of #209)
- [x] [BC Break] Add `<SimpleShowLayout>` component to `<Show>` view

Closes #216

tabbed-form

### CODE REVIEW

File: docs/CreateEdit.md [hunk 1]
```diff
@@ -5,20 +5,81 @@ title: "The Create and Edit Views"
...
+        </SimpleForm>
+    </Edit>
+);
+```
+
+That's enough to display the post edit form:
+
+![post edition form](./img/post-edition.png)
+
+**Tip**: You might find it cumbersome to repeat the same input components for both the `<Create>` and the `<Edit>` view. In practice, these two views almost never have exactly the same form inputs. For instance, in the previous snippet, the `<Edit>` views shows related comments to the current post, which makes no sense for a new post. Having two separate sets of input components for the two view is therefore a deliberate choice. However, if you have the same set of input components, export them as a custom from component to avoid repetition.
```
Comment (jpetitcolas) [inline docs/CreateEdit.md]: s/from component to/form component to/

File: docs/CreateEdit.md [hunk 2]
```diff
@@ -79,9 +140,80 @@ export const PostEdit = (props) => (
...
+            <RichTextInput source="body" />
+            <NumberInput source="nb_views" />
+        </SimpleForm>
+    </Create>
+);
+```
+
+## The `<TabbedForm>` component
+
+Just like `<SimpleForm>`, `<TabbedForm>` the `record` prop, renders the actual form, and handles form validation on submit. However, the `<TabbedForm>` component renders inputs grouped by tab. The tabs are set by using `<FormTab>` components, which expect a `label` and an `icon` prop.
```
Comment (jpetitcolas) [inline docs/CreateEdit.md]: s/Form>`/Form>` receives/

File: docs/CreateEdit.md (hunk 1 above)
Comment (djhi) [inline docs/CreateEdit.md]: s/from component to/component to/

File: docs/CreateEdit.md (hunk 2 above)
Comment (djhi) [inline docs/CreateEdit.md]: s/`<TabbedForm>` the/`<TabbedForm>` receives the/

File: docs/Inputs.md [hunk 3]
```diff
@@ -208,28 +198,35 @@ import { DisabledInput } from 'admin-on-rest/lib/mui';
...
+        <SimpleForm>
+            <TextField source="title" /> {/* NOT EDITABLE */}
+            <LongTextInput source="body" />
+        </SimpleForm>
     </Edit>
 );
 ```
 
-**Tip**: You can even use a component of your own, provided it accepts `record` and `label` props:
+**Tip**: You can even use a component of your own, provided it accepts `record` prop:
```
Comment (djhi) [inline docs/Inputs.md]: s/it accepts/it accepts a/

File: docs/CreateEdit.md (hunk 1 above)
Comment (Kmaschta) [inline docs/CreateEdit.md]: s/from component to/component to/

**Pull request merged**